### PR TITLE
Added tool for splitting a configurations set into a training set and a test set.

### PR DESCRIPTION
**Pull request by EthanBullVulpe**
The module configurations_splitter.py has been added, it contains a method used to divided a configurations set into a training set and a test set using furthest point sampling.

the method is generalized to be able to use any MolecularDdescriptor() to perform the sampling. There is currently only the RMSDDescriptor, which describes the difference between two molecules based on the rmsd between the locations of the atoms inside them.

### CODE REVIEW

File: qm_mb_energy_calculator/src/molecule.py [hunk 1]
```diff
@@ -781,6 +783,49 @@ def rmsd(self, other):
         # compute rmsd as sqrt of mean squared distance
         return math.sqrt(squared_distance / self.get_num_atoms())
 
+    def distancermsd(self, other_molecule):
```
Comment (EthanBullVulpe) [inline qm_mb_energy_calculator/src/molecule.py]: is it clear how the two rmsd functions differ based on the comments?

File: configuration_generator/src/configurations_splitter.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    # parse the configurations into a list of [id, molecule]
+    molecules = list(enumerate(molecule_parser.xyz_to_molecules(configurations_path, settings_reader.SettingsReader(settings_file))))
+
+    # move all molecules to their center of mass and rotate them on their principal axes
+    for index, molecule in molecules:
+        molecule.move_to_center_of_mass()
+        molecule.rotate_on_principal_axes()
+
+    # construct a matrix where the item at index [i, k] is the difference between the ith molecule and the kth molecule
+    # it starts empty, and will be filled in
```
Comment (EthanBullVulpe) [inline configuration_generator/src/configurations_splitter.py]: we do this, so we only have to compute each difference once at maximum.

Comment (EthanBullVulpe) [reply configuration_generator/src/configurations_splitter.py]: While it makes the code a little more complicated than simply calculating the difference each time it is needed, this allows the configuration_splitter to be MUCH faster.

File: configuration_generator/src/configurations_splitter.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    difference_matrix = [[None for k in range(i + 1)] for i in range(len(molecules) - 1)]
```
Comment (EthanBullVulpe) [inline configuration_generator/src/configurations_splitter.py]: We only construct the bottom half of the matrix, because the rest would be redundant information.

File: configuration_generator/src/configurations_splitter.py (hunk 2 above)
Comment (agoetz) [reply configuration_generator/src/configurations_splitter.py]: This is a good idea. We may have to be careful because it is not scalable to very large data sets. Assuming 64bit floats, this requires storage of 8*N*(N-1)/2 byte of storage. The water 2b training set contains 42k entries, which means 8*42*21 = 7GB storage. A training set with 100k entries requires already 78GB storage.

File: configuration_generator/src/configurations_splitter.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+class MolecularDescriptor():
+    def difference(molecule1, molecule2):
+        raise NotImplementedError
+
+class RMSDDescriptor(MolecularDescriptor):
+    def difference(self, molecule1, molecule2):
+        return molecule1.rmsd(molecule2)
+
+class RMSDDistanceDescriptor(MolecularDescriptor):
+    def difference(self, molecule1, molecul2):
```
Comment (agoetz) [inline configuration_generator/src/configurations_splitter.py]: Typo, should be `molecule2`

File: qm_mb_energy_calculator/src/molecule.py (hunk 1 above)
Comment (agoetz) [reply qm_mb_energy_calculator/src/molecule.py]: I understand it. It's sort of hard to put in words. Maybe "Computes the RMSD of intramolecular interatomic distances between two molecules".

The D in RMSD stands for difference, so it is superfluous. It is RMSD or RMS difference.

**Comment by agoetz** [comment]
@EthanBullVulpe Great that you implemented furthest point sampling. This will be very useful. There's just one typo. Let me know when it is fixed so I can merge or just merge yourself.

File: configuration_generator/src/configurations_splitter.py (hunk 2 above)
Comment (EthanBullVulpe) [reply configuration_generator/src/configurations_splitter.py]: This is true, but this furthest point sampling would be grossly inefficient for a training set of that size anyway. There might be a way to do furthest point sampling with a more efficient runtime than O(N^2), but I haven't thought about it
